contents/newsletter/feature-flag-mistakes.md

```diff
@@ -0,0 +1,178 @@
+  return pricingStrategy.calculatePrice(product);
+}
+
+// Flag logic separated into configuration/initialization
+const pricingStrategy = posthog.isFeatureEnabled('new_pricing_algorithm') 
+  ? new NewPricingStrategy()
+  : new StandardPricingStrategy();
+```
+
+An ideal structure would use the flag in as few places as possible and abstract as much as possible behind a single flag. This creates a clear separation and isolation of flag-related code which limits complexity and improves maintainability.
+
+Good goals for avoiding this mistake are only using the feature flag in a single spot and making it easy to remove. Accomplishing both reveals your code is structured well. 
+
+## 3. Being surprised when flags are turned off
+
+The core of a feature flag's value to developers is the ability to turn them off. They need to be ready for this at any point, which isn't always the case.
+
+![Shocked by feature flag](https://res.cloudinary.com/dmukukwp6/image/upload/turnoff_941966283f.png)
+
+As an example, a [commenter on Hacker News](https://news.ycombinator.com/item?id=24550442) shared that they had two major incidents in two years due to their feature flag service failing to return flag states and old flags unintentionally being set to false.
+
+This hits close to home. Last year, we had an [incident](https://github.com/PostHog/post-mortems/blob/main/investigations/2024-02-28-decide%20is%20down.md) with our feature flag service that caused flags to fallback to empty responses. On top of this, some requests had exceeding long timeouts. The combination of these caused some of our customer's applications to crash. 
+
+Of course, after fixing this, we learned and made changes like enabling customized timeouts, lowering the default timeout from 10 to 3 seconds, and moving flag requests out of our Django monolith. Regardless, it shows you a situation where flags can have an unexpected state and cause issues.
+
+Preventing this issue is a two step process: 
+
+1. Test each potential state of the flag. This includes a flag returning `null`. 
+2. Fall back to code that works. The default should always be that your app works. 
+
+## 4. Holding on to stale flags forever
+
+As the saying goes: "If you love something, set it free." The same is true of feature flags. 
+
+The opposite of this is [Knight Capital Group](https://www.henricodolfing.com/2019/06/project-failure-case-study-knight-capital.html), a financial services firm that was *formerly* the largest U.S. equities trader on the NYSE.
+
+To support a new private market being launched by the NYSE, their team needed to build and deploy an updated trade execution system, responsible for thousands of orders per second, in a month. 
+
+After rushing to meet an August 1st, 2012 deadline, their team shipped an update replacing code in a part of their system unused since 2003. This old code wasn't meant to be used in production, but was left behind a feature flag in their codebase anyways.
```

Comment:
```suggestion
After rushing to meet an August 1st, 2012 deadline, they shipped an update replacing code in a part of their system unused since 2003. This old code wasn't meant to be used in production, but was left behind a feature flag in their codebase anyways.
```